text/0000-custom-dst.md

```diff
@@ -0,0 +1,476 @@
+    type Metadata = ();
+
+    fn size_of_val(&self) -> usize { size_of::<T>() }
+    fn align_of_val(&self) -> usize { align_of::<T>() }
+}
+```
+
+If you have a type which you would like to be unsized,
+you can implement this trait for your type!
+
+```rust
+#[repr(C)]
+struct CStr([c_char; 0]);
+
+unsafe impl DynamicallySized for CStr {
+    type Metadata = ();
+
+    fn size_of_val(&self) -> usize { strlen(&self.0 as *const c_char) }
+    fn align_of_val(&self) -> usize { 1 }
+}
+```
+
+and automatically, your type will not implement `Sized`.
+
+The existing `DynamicallySized` types will continue to work;
+if one writes a `DynamicallySized` type `T`,
+and then wraps `T` into a struct, they'll get the obvious semantics.
+
+```rust
+struct Foo {
+    x: usize,
+    y: CStr,
+}
+
+// size_of_val(&foo) returns size_of::<usize>() + size_of_val(&foo.y)
+// same with align_of_val
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+In addition to the explanation given above,
+we will also introduce three functions into the standard library,
+in `core::raw`, which allow you to create and destructure these
+pointers to `DynamicallySized` types:
+
+```rust
+mod core::raw {
+    pub fn from_raw_parts<T: DynamicallySized>(
+        ptr: *const (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *const T;
+
+    pub fn from_raw_parts_mut<T: DynamicallySized>(
+        ptr: *mut (),
+        meta: <T as DynamicallySized>::Metadata,
+    ) -> *mut T;
+
+    pub fn metadata<T: DynamicallySized>(
+        ptr: *const T,
+    ) -> <T as DynamicallySized>::Metadata;
+}
+```
+
+and we will introduce two functions into `core::mem`,
+to help people write types with Flexible Array Members:
+
+```rust
+mod core::mem {
+    pub fn size_of_header<T: ?Sized>() -> usize;
+    pub fn align_of_header<T: ?Sized>() -> usize;
+}
+```
+These functions return the size and alignment of the header of a type;
+or, the minimum possible size and alignment, in other words.
+For existing `Sized` types, they are equivalent to `size_of` and `align_of`,
+and for existing DSTs,
+
+```rust
+assert_eq!(size_of_header::<[T]>(), 0);
+assert_eq!(align_of_header::<[T]>(), align_of::<T>());
+assert_eq!(size_of_header::<dyn Trait>(), 0);
+assert_eq!(align_of_header::<dyn Trait>(), 1);
+
+// on 64-bit
+struct RcBox<T: ?Sized> {
+  strong: Cell<usize>,
+  weak: Cell<usize>,
+  value: T,
+}
+assert_eq!(size_of_header::<RcBox<dyn Trait>>(), 16);
+assert_eq!(align_of_header::<RcBox<dyn Trait>>(), 8);
+```
+
+Note that this is a minimum - this means that for `extern type`s,
+they return `0` and `1` respectively.
+
+Notes:
+  - names of the above functions should be bikeshed
+  - `extern type`s do not implement `DynamicallySized`, although in theory one
+    could choose to implement the trait for them
+    (that usecase is not supported by this RFC).
+  - `T: DynamicallySized` bounds imply a `T: ?Sized` bound.
```

Comment:
N.B. `Type: Trait` is not a bound (because `Trait + 'static` is a bound...), it is a *constraint*.

Reply:
What's the distinction?

Reply:
So a `trait Foo { ... }` has an implicit first parameter we call `Self` which is not present in Haskell; This means that when you say something like `MyTrait`, it has the kind `bound = k1 -> constraint` where `k1 : type | lifetime ;`. The operator `:` in this context can then be understood as something of kind `type -> bound -> constraint`. Notice that when you say `Show Int =>` in Haskell, `Show :: * -> Constraint` (a "bound") but `Show Int :: Constraint`.

Reply:
ah, I see!